[uniqly] iP

uniqly opened the pull request:
# EKuD 🙃

> "To be, or not to be, that is the question" -- Hamlet ([source](https://en.wikipedia.org/wiki/To_be,_or_not_to_be))

~~Duke~~EKuD is out for _**REVENGE**_ against CS2103 students who _keep messing with his code_, and **HE... IS.. FURIOUS**.

With his _new_ makeover, EKuD has:

- 100% more angst
- 100% more sass
- 100% more intolerance of your [REDACTED]

EKuD invites you try him out! Just do the following:

1. Download `EKuD.jar` from [here](https://github.com/uniqly/ip/releases/tag/Level-10)
1. Double click the file
1. ~~Solve world hunger~~
1. And BOOM! you are good to go 🥳 

Features:

- [x] Manage 3 different kinds of tasks
- [x] Text based CLI UI
- [x] Graphical based UI  
- [ ] Free million dollars

## CLI Mode
EKuD runs in GUI mode by default, but it also supports a legacy CLI interface:

```java
public class Launcher {
    public static void main(String[] args) {
        if (args.length == 1 && args[0].equals("--text")) {
            Ekud.runTextMode();
            System.exit(0);
        } else if (args.length > 0) {
            System.out.println("Invalid flag, defaulting to GUI mode.\nUse '--text' for CLI mode.");
        }
        Application.launch(Main.class, args);
    }
}
```

Just run `java -jar EKuD.jar --text` in your command line or shell!!!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

uniqly committed:
Ekud.java: Rename Duke.java, Change greeting and exit

uniqly committed:
Ekud.java: Add greeting and exit static methods

uniqly committed:
Ekud.java: Add identation to all messages

uniqly committed:
Ekud.java: Add echo requirement to EKuD

uniqly committed:
Ekud.java: Change static methods to class methods

uniqly committed:
Ekud.java: Add "Add, List" requirements

uniqly committed:
Add Task.java

uniqly committed:
Task.java: Add public modifier to constructor

uniqly committed:
Ekud.java: Change list to array of Task, Rename list to tasks

uniqly committed:
Ekud.java: Add markList() and unmarkList() methods

uniqly committed:
Add TodoTask.java

uniqly committed:
Add DeadlineTask.java

uniqly committed:
Add EventTask.java

uniqly committed:
Ekud.java: Add token parser for user inputs

uniqly committed:
Ekud.java: Add parsing of todo, deadline, event task

uniqly committed:
Task.java: Add isDone() method

uniqly committed:
Add TaskList.java

uniqly committed:
Ekud.java: Change tasks to TaskList

uniqly committed:
bug fix: Fix task count not increasing in list command

uniqly committed:
Update TextUiTesting inputs

uniqly committed:
TaskList.java: Change tasks to ArrayList

uniqly committed:
Ekud.java: Remove echo method

uniqly committed:
Ekud.java: Add error message for unknown command

uniqly committed:
Task.java: Change to abstract class

uniqly committed:
Add FormatPrinter.java

uniqly committed:
Ekud.java: Replace echo with FormatPrinter printIdent

uniqly committed:
Add TaskArgumentMissingException.java

uniqly committed:
Add error message for empty Task description

uniqly committed:
Add TaskListIndexOutOfBoundsException.java

uniqly committed:
Add EkudWrongInputFormatException.java

uniqly committed:
TaskList.java: Add checkTask() method

uniqly committed:
Add error messages for invalid list index

uniqly committed:
Add missing args error message for Deadline and Event task

uniqly committed:
bugfix: Fix typo in mark list message

uniqly committed:
bugfix: Fix missing error check for unmark list

uniqly committed:
TaskList.java: Add getCount() and removeTask() methods

uniqly committed:
Add delete task command

uniqly committed:
TaskList.java: Add getCount() and isAllComplete() methods

uniqly committed:
Add completion status for task related commands

uniqly committed:
Update Text-UI-Test test case

uniqly committed:
TaskList.java: Add final modifier to tasks

uniqly committed:
Ekud.java: Close Scanner in main() method

uniqly committed:
Add method to get save string for all Tasks

uniqly committed:
Task.java: Add factory method for taskSaveString to Task

uniqly committed:
Add reading and creating of save file

uniqly committed:
Add ability to clean up invalid task data files

uniqly committed:
Add task data update after add/delete/(un)mark

uniqly committed:
Merge branch 'branch-Level-7'

uniqly committed:
Change DeadlineTask to store LocalDateTime

uniqly committed:
Change EventTask to store LocalDateTime

uniqly committed:
Merge branch 'branch-Level-8'

uniqly committed:
Refactor Ekud into more OOP classes

uniqly committed:
Remove obsolete classes

uniqly committed:
Merge branch 'branch-A-MoreOOP'

uniqly committed:
Organize files into ekud package

uniqly committed:
bugfix: Fix extra decrement of int input by parser

uniqly committed:
Merge branch 'branch-A-Packages'

uniqly committed:
Merge branch 'add-gradle-support'

uniqly committed:
Configure gradle to use Main.java

uniqly committed:
Add unit testing for Task components

uniqly committed:
DeadlineTask.java: Fix wrong tag in empty description string

uniqly committed:
Merge branch 'branch-A-JUnit'

uniqly committed:
bugfix: Fix save file not being generated

uniqly committed:
Rename gradle archiveBaseName to EKuD

uniqly committed:
Add JavaDoc to all classes in ekud package

uniqly committed:
Remove Main and move driver code to Ekud

uniqly committed:
Fix outstanding coding style violations

uniqly committed:
Add find keyword command

uniqly committed:
Merge branch 'branch-A-JavaDoc'

uniqly committed:
Merge branch 'branch-A-CodingStandard'

uniqly committed:
Merge branch 'branch-Level-9'

uniqly committed:
Add checkstyle xml files

uniqly committed:
Reconfigure build.gradle with checkstyle

uniqly committed:
Fix checkstyle errors for import statements

uniqly committed:
Command.Type class: Fix static field before static method

uniqly committed:
Fix checkstyle errors for trailing whitespace

uniqly committed:
Fix missing final modifier for constants

uniqly committed:
Add checkstyle suppression for arrow-head switch case

uniqly committed:
Fix checkstyle errors for test files

uniqly committed:
Merge branch 'branch-A-CheckStyle'

uniqly committed:
Move Ui.java to ekud.ui package

uniqly committed:
Add skeleton GUI

uniqly committed:
Refactor Ui class to abstract class with TextUi subclass

uniqly committed:
Integrate EKuD with GUI

uniqly committed:
Make GUI window auto-resizing

uniqly committed:
Add css styling to GUI

uniqly committed:
DialogBox.java: Fix import statements order

uniqly committed:
Merge branch 'branch-Level-10'

noth-k commented on src/main/java/ekud/Ekud.java:
```diff
@@ -0,0 +1,133 @@
+     * Loads task data from storage.
+     */
+    public void loadTasks() {
+        ui.addToBuffer("Before we start, let me try to find your tasks!!");
```
Could these string literals be constants, which could make it easier to update it in the future?

noth-k commented on src/main/java/ekud/Ekud.java:
```diff
@@ -0,0 +1,133 @@
+package ekud;
+
+import java.util.Scanner;
```
I like how the imports are grouped up :)

noth-k commented on src/main/java/ekud/commands/Command.java:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        /**
+         * Converts a given {@link String} into a {@link Type}.
```
Could the comment for getType method could be clearer? Maybe instead of "Converts a given {@link String} into a {@link Type}.", it might be more descriptive to say "Returns the Command Type associated with the given input string."?

noth-k commented on src/main/java/ekud/components/Parser.java:
```diff
@@ -0,0 +1,94 @@
+        tokenMap.put("command", words[0]);
+        // get params
+        String currToken = "argument";
+        for (int i = 1; i < words.length; i++) {
```
Nice job surrounding the Operators with the space character.

noth-k commented on src/main/java/ekud/components/Parser.java:
```diff
@@ -0,0 +1,94 @@
+            case LIST -> new ListCommand();
+            case FIND -> new FindCommand(argument);
+            case EXIT -> new ExitCommand();
+        };
```
nice indentation here!

noth-k reviewed (commented):
Overall nice one! The JavaDocs were descriptive and understandable and coding standards were followed

yikjunxian commented on src/main/java/ekud/task/Task.java:
```diff
@@ -0,0 +1,175 @@
+     *
+     * @return Error message String.
+     */
+    public abstract String getEmptyDescriptionErrorMessage();
```
Good use of abstract method to ensure that the specific tasks have relevant error messages!

yikjunxian commented on src/main/java/ekud/task/Task.java:
```diff
@@ -0,0 +1,175 @@
+import ekud.ui.Ui;
+
+/**
+ * Represents a task. Each task minimally has a description and completion status.
```
Consider adding to the JavaDoc that Task is an abstract class and not meant to be instantiated.

yikjunxian commented on src/main/java/ekud/task/Task.java:
```diff
@@ -0,0 +1,175 @@
+     * @throws EkudException If description is null or empty.
+     */
+    public Task(String description) throws EkudException {
+        if (description == null || description.isEmpty()) {
```
Consider checking if users input only whitespaces as task description using description.trim().isEmpty().

yikjunxian commented on src/main/java/ekud/task/Task.java:
```diff
@@ -0,0 +1,175 @@
+    public static Task getTaskFromSave(String taskSaveString, Ui ui) {
+        try {
+            taskSaveString = taskSaveString.trim(); // removes extra new lines
+            String[] args = taskSaveString.split("(\\s\\|\\s)"); // splits string along < | > delimiter
```
Nice use of delimiters to extract relevant components of the string.

yikjunxian commented:
LGTM. Just a few nits to fix.

yikjunxian reviewed (approved):
LGTM. Just a few nits to fix.

uniqly committed:
Add assertions to main codebase

Current codebase does not use assertions.

Assertions are used to define assumptions about the program state so
that run time can verify them. An assertion failure indicates a
possible bug in the code. Without assertions, errors that result in
the assumptions being violated will be propogated down the code
resulting in further potential bugs.

Let's,
* check for any assumptions used within the codebase
* define the assumption using the `assert` keyword

uniqly committed:
bugFix: Remove assertion for checking if temp file is deleted

uniqly committed:
Change Add/Mark/Unmark Commands to take 0 based index

uniqly committed:
Ui.java: Add addFormattedToBuffer() method

uniqly committed:
commands package: Remove magic literals

Classes in `ekud.commands` currently have many magic literals related to
outputs to the UI when `Command#execute` is called.

Magic literals make the codebase more unreadable as the literals used do
not explain what they are used for or what they mean.

Let's aim to code improve readability by replacing instances of magic
literals with named alternatives.

uniqly committed:
Parser.java: Replace magic literals

uniqly committed:
Replace existing magic literals

uniqly committed:
TodoTask.java: Fix toString() not printing task status

uniqly committed:
TaskList.java: Replace complicated boolean in constructor

uniqly committed:
Parser.java: Improve readability of tokenize() method

uniqly committed:
TaskList.java: Remove unused methods

uniqly committed:
Update JavaDoc to be less ambigious for abstract classes

uniqly committed:
Storage.java: Refactor to use Streams for reading data

uniqly committed:
Merge pull request #1 from uniqly/branch-A-Assertions

Add assertions to codebase

uniqly committed:
Merge branch 'master' into branch-A-CodeQuality

uniqly committed:
Merge branch 'master' into branch-A-Streams

uniqly committed:
Merge pull request #2 from uniqly/branch-A-CodeQuality

Update code base to reflect better code quality

uniqly committed:
Merge branch 'master' into branch-A-Streams

uniqly committed:
Merge pull request #3 from uniqly/branch-A-Streams

Use streams within code base

uniqly committed:
Add CI workflow for gradle

uniqly committed:
bugfix: Remove extra space in TodoTask#toString() output

uniqly committed:
Suppress checkstyleTest warnings for lambda expressions

uniqly committed:
Fix text-ui-test

uniqly committed:
bugfix: Fix text-ui-test not accounting for data subfolder

uniqly committed:
Merge pull request #5 from uniqly/branch-A-CI

Add CI workflow integeration

uniqly committed:
Add ability to give tasks a priority

uniqly committed:
bugfix: Fix typo in invalid task type error message

uniqly committed:
Update text-ui-test for task priority

uniqly committed:
Update unit tests for task priority

uniqly committed:
Add comparators for Tasks

uniqly committed:
Add the ability to sort tasks

uniqly committed:
Update text-ui-test for sorting tasks

uniqly committed:
Merge pull request #6 from uniqly/branch-BCD-Extension

Implement C-Priority and C-Sort

uniqly committed:
Adjust colors and font family in GUI

uniqly committed:
bug fix: Fix output showing `ekud.task` instead of `task`

uniqly committed:
EventTask: Add check for if from date is before to date

uniqly committed:
bug fix: Fix negative index not being handled correctly

uniqly committed:
Task class: Add check for completion before marking/unmarking

uniqly committed:
FindCommand.java: Fix typo in error message

uniqly committed:
Task class: Add error for if attempting to set same priority

uniqly committed:
bugfix: Fix input dates with more days than days in month

uniqly committed:
Update text-ui-test for new error checks

uniqly committed:
Add error hanlding for duplicate parameters

uniqly committed:
Merge pull request #7 from uniqly/branch-A-MoreErrorHandling

Add more error handling

uniqly committed:
Add user guide

uniqly committed:
Replace incompatible alerts syntax in user guide

uniqly committed:
Fix typo in user guide at QNA#Saving section

uniqly committed:
Merge pull request #8 from uniqly/branch-A-UserGuide

Create user guide
